File: app/backend/Events/EventsAPI/Services/Implementations/EventsService.cs
using EventsAPI.Data;
using EventsAPI.Models;
using EventsAPI.Models.DTOS;
using EventsAPI.Services.Contracts;
using Microsoft.EntityFrameworkCore;

namespace EventsAPI.Services.Implementations
{
    public class EventsService : IEventsService
    {
        private const int MIN_PAGE_NUMBER = 1;
        private const int MIN_PAGE_SIZE = 5;
        private readonly EventsDbContext _dbContext;

        public EventsService(EventsDbContext dbContext)
        {
            _dbContext = dbContext;
        }

        public async Task<int> CreateEvent(NewEvent createEvent)
        {
            // TODO: add validation
            var newEvent = new Event
            {
                Title = createEvent.Title,
                Description = createEvent.Description,
                Location = createEvent.Location,
                Image = createEvent.Image,
                Featured = createEvent.Featured,
                EventDates = createEvent.EventDates.Select(e => new EventDate
                {
                    Date = e.Date,
                    Price = e.Price
                }).ToList()
            };

            await _dbContext.Events.AddAsync(newEvent);
            await _dbContext.SaveChangesAsync();

            return newEvent.Id; // TODO: investigate whether we should/can return the id or not
        }

        public async Task<EventDetailsResult> GetEvent(int id)
        {
            var result = await _dbContext.Events.
                Include(e => e.EventDates)
                .FirstOrDefaultAsync(e => e.Id == id);

            if (result == null)
            {
                return null;
            }

            // TODO: use automapper?
            return new EventDetailsResult
            {
                Id = result.Id,
                Title = result.Title,
                Description = result.Description,
                Location = result.Location,
                Image = result.Image,
                Featured = result.Featured,
                EventDates = result.EventDates.Select(e => new EventDetailsDateResult
                {
                    Id = e.Id,
                    Date = e.Date,
                    Price = e.Price
                }).ToList()
            };
        }

        public async Task<IEnumerable<EventDateResult>> GetEventDates(bool featured, int pageNumber, int pageSize)
        {
            pageNumber = pageNumber < MIN_PAGE_NUMBER ? MIN_PAGE_NUMBER : pageNumber;
            pageSize = pageSize < MIN_PAGE_SIZE ? MIN_PAGE_SIZE : pageSize;

            return await _dbContext.EventDates
                .Where(e => e.Event.Featured == featured)
                .OrderBy(e => e.Date)
                .Skip((pageNumber - 1) * pageSize)
                .Take(pageSize)
                .Select(e => new EventDateResult
                {
                    EventId = e.Event.Id,
                    EventDateId = e.Id,
                    Title = e.Event.Title,
                    Description = e.Event.Description,
                    Image = e.Event.Image,
                    Featured = e.Event.Featured,
                    Date = e.Date
                })
            .ToListAsync();
        }
    }
}

File: app/frontend/events-app/src/EventsService.js
const EVENTS_API_URL = 'https://localhost:7297/api/Events';

class EventsService {
    getEvents(featured, pageNumber, pageSize) {
        let url = EVENTS_API_URL;

        url += `?featured=${featured}`;

        if (pageNumber) {
            url += `&pageNumber=${pageNumber}`;
        }

        if (pageSize) {
            url += `&pageSize=${pageSize}`;
        }

        console.log(`Calling ${url} ...`);

        return fetch(url);
    }

    getEvent(id) {
        return fetch(`${EVENTS_API_URL}/${id}`);
    }

    createEvent(newEvent) {
        const formData = new FormData();
    
        formData.append('title', newEvent.title);
        formData.append('description', newEvent.description);
        formData.append('location', newEvent.location);
        formData.append('featured', newEvent.featured);
        formData.append('image', newEvent.image);

        // for (const date of newEvent.dates) {
        //     formData.append('dates', date);
        // }

        // TODO: for multiple imgs
        // for (const img of data.imgs) {
        //     formData.append('img', img);
        // }

        return fetch(EVENTS_API_URL, {
            method: 'POST',
            cache: 'no-cache',
            body: formData
        });
    }
}


export default new EventsService();